### PR TITLE
Allow setting example paths and queries for DocService

### LINKED ISSUE #2195: Provide a way to set example query string
Currently, we can only set example headers and requests in `DocService` for annotated services. A user might want to set an example query or example path as well.

### PR DESCRIPTION
**Pull request by ikhoon**
Motivation:

Currently, a user cannot set example paths and queries. #2195
The endpoint path or queries are only set automatically when
opening a URL that contains `endpoint_path` or `queries` as params.

Modifications:

- Add `examplePaths` to `DocService` and `DocServiceBuilder`
- Add `exampleQueries` to `DocService` and `DocServiceBuilder`
- Show `examplePath` if the `endpoint_path` is null and the `examplePath` is not null
- Show `exampleQuery` if the `queries` is null and the `exampleQuery` is not null
- Make `ENDPOINT PATH` input visible with read-only if a path is exact type 

Result:
You can now set an example path and queries for `DocService`
Fixes #2195

### CODE REVIEW

**Comment by ikhoon** [comment]
A screenshot of example paths and queries
image

File: core/src/main/java/com/linecorp/armeria/server/docs/DocService.java [hunk 1]
```diff
@@ -108,14 +110,17 @@ public static DocServiceBuilder builder() {
      * Creates a new instance.
      */
     public DocService() {
-        this(ImmutableMap.of(), ImmutableMap.of(), ImmutableList.of(), DocServiceBuilder.ALL_SERVICES);
+        this(ImmutableMap.of(), ImmutableMap.of(), ImmutableMap.of(),
```
Comment (anuraaga) [inline core/src/main/java/com/linecorp/armeria/server/docs/DocService.java]: This was already a bit crazy but now even more so probably needs param comments e.g., `/* exampleHttpHeaders= */ ImmutableMap.of()`

Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/server/docs/DocService.java]: Let me add param comments.🤣🙈

File: docs-client/src/containers/MethodPage/EndpointPath.tsx [hunk 2]
```diff
@@ -37,15 +41,25 @@ const EndpointPath: React.FunctionComponent<Props> = (props) => (
...
-          rows={1}
-          value={props.endpointPath}
+          value={props.additionalPath}
           placeholder={endpointPathPlaceHolder}
-          onChange={props.onEndpointPathChange}
+          onChange={props.onPathFormChange}
           inputProps={{
-            className: 'code',
+            readOnly: !props.editable,
+            className: 'code ',
```
Comment (hexoul) [inline docs-client/src/containers/MethodPage/EndpointPath.tsx]: Out of curiosity: is there special meaning for a space after `code`?

Comment (ikhoon) [reply docs-client/src/containers/MethodPage/EndpointPath.tsx]: Oops... it's my mistake. thanks for catching it!

File: core/src/main/java/com/linecorp/armeria/server/docs/DocServiceBuilder.java [hunk 3]
```diff
@@ -183,6 +185,90 @@ private DocServiceBuilder exampleHttpHeaders0(String serviceName, String methodN
...
+            this.examplePaths.computeIfAbsent(serviceName, unused -> ArrayListMultimap.create())
+                             .put(methodName, examplePath);
+        }
+        return this;
+    }
+
+    /**
+     * Adds the specified example queries for the method with the specified service and method name.
+     */
+    public DocServiceBuilder exampleQueries(Class<?> serviceType, String methodName, String... queries) {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/docs/DocServiceBuilder.java]: `queries` -> `queryStrings` ?

File: core/src/main/java/com/linecorp/armeria/server/docs/MethodInfo.java [hunk 4]
```diff
@@ -108,6 +114,8 @@ public MethodInfo(String name,
         this.exampleHttpHeaders = ImmutableList.copyOf(requireNonNull(exampleHttpHeaders,
                                                                       "exampleHttpHeaders"));
         this.exampleRequests = ImmutableList.copyOf(requireNonNull(exampleRequests, "exampleRequests"));
+        this.examplePaths = ImmutableList.copyOf(requireNonNull(examplePaths, "examplePaths"));
+        this.exampleQueries = ImmutableList.copyOf(requireNonNull(exampleQueries, "exampleQueries"));
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/docs/MethodInfo.java]: How about validating all paths and query strings, using `PathAndQuery` and `QueryParams.fromQueryString()`?

File: core/src/main/java/com/linecorp/armeria/server/docs/DocServiceBuilder.java [hunk 5]
```diff
@@ -183,6 +186,95 @@ private DocServiceBuilder exampleHttpHeaders0(String serviceName, String methodN
...
+     */
+    public DocServiceBuilder exampleQueries(Class<?> serviceType, String methodName, String... queryStrings) {
+        requireNonNull(serviceType, "serviceType");
+        return exampleQueries(serviceType.getName(), methodName, queryStrings);
+    }
+
+    /**
+     * Adds the specified example query strings for the method with the specified service and method name.
+     */
+    public DocServiceBuilder exampleQueries(Class<?> serviceType, String methodName, Iterable<String> queries) {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/docs/DocServiceBuilder.java]: `queries` -> `queryStrings`

File: core/src/main/java/com/linecorp/armeria/server/docs/DocServiceBuilder.java [hunk 6]
```diff
@@ -183,6 +186,95 @@ private DocServiceBuilder exampleHttpHeaders0(String serviceName, String methodN
...
+     */
+    public DocServiceBuilder exampleQueries(String serviceName, String methodName, String... queryStrings) {
+        requireNonNull(queryStrings, "queryStrings");
+        return exampleQueries(serviceName, methodName, ImmutableList.copyOf(queryStrings));
+    }
+
+    /**
+     * Adds the specified example query strings for the method with the specified service and method name.
+     */
+    public DocServiceBuilder exampleQueries(String serviceName, String methodName, Iterable<String> queries) {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/docs/DocServiceBuilder.java]: `queries` -> `queryStrings`

File: core/src/main/java/com/linecorp/armeria/server/docs/DocServiceBuilder.java [hunk 7]
```diff
@@ -183,6 +186,95 @@ private DocServiceBuilder exampleHttpHeaders0(String serviceName, String methodN
         return this;
     }
 
+    /**
+     * Adds the specified example paths for the method with the specified service and method name.
+     */
+    public DocServiceBuilder examplePaths(Class<?> serviceType, String methodName, String... examplePaths) {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/docs/DocServiceBuilder.java]: `paths`? (like `queryStrings` below)

File: core/src/main/java/com/linecorp/armeria/server/docs/DocServiceBuilder.java [hunk 8]
```diff
@@ -183,6 +186,95 @@ private DocServiceBuilder exampleHttpHeaders0(String serviceName, String methodN
...
+        requireNonNull(serviceType, "serviceType");
+        return examplePaths(serviceType.getName(), methodName, examplePaths);
+    }
+
+    /**
+     * Adds the specified example paths for the method with the specified service and method name.
+     */
+    public DocServiceBuilder examplePaths(Class<?> serviceType, String methodName,
+                                          Iterable<String> examplePaths) {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/docs/DocServiceBuilder.java]: `paths`? (like `queryStrings` below)

File: core/src/main/java/com/linecorp/armeria/server/docs/DocServiceBuilder.java [hunk 9]
```diff
@@ -183,6 +186,95 @@ private DocServiceBuilder exampleHttpHeaders0(String serviceName, String methodN
...
+        requireNonNull(serviceType, "serviceType");
+        return examplePaths(serviceType.getName(), methodName, examplePaths);
+    }
+
+    /**
+     * Adds the specified example paths for the method with the specified service and method name.
+     */
+    public DocServiceBuilder examplePaths(String serviceName, String methodName, String... examplePaths) {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/docs/DocServiceBuilder.java]: `paths`? (like `queryStrings` below)

File: core/src/main/java/com/linecorp/armeria/server/docs/DocServiceBuilder.java [hunk 10]
```diff
@@ -183,6 +186,95 @@ private DocServiceBuilder exampleHttpHeaders0(String serviceName, String methodN
...
+        requireNonNull(examplePaths, "examplePaths");
+        return examplePaths(serviceName, methodName, ImmutableList.copyOf(examplePaths));
+    }
+
+    /**
+     * Adds the specified example paths for the method with the specified service and method name.
+     */
+    public DocServiceBuilder examplePaths(String serviceName, String methodName,
+                                          Iterable<String> examplePaths) {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/docs/DocServiceBuilder.java]: `paths`? (like `queryStrings` below)

File: core/src/main/java/com/linecorp/armeria/server/docs/MethodInfo.java (hunk 4 above)
Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/server/docs/MethodInfo.java]: This is not addressed yet.

File: core/src/main/java/com/linecorp/armeria/server/docs/MethodInfo.java [hunk 11]
```diff
@@ -117,16 +118,25 @@ public MethodInfo(String name,
...
 
-        this.examplePaths = ImmutableList.copyOf(requireNonNull(examplePaths, "examplePaths"));
-        this.examplePaths.forEach(path -> checkArgument(PathAndQuery.parse(path) != null,
-                                                        "examplePaths contains an invalid path: {}", path));
+        requireNonNull(examplePaths, "examplePaths");
+        final ImmutableList.Builder<String> examplePathsBuilder =
+                ImmutableList.builderWithExpectedSize(Iterables.size(examplePaths));
+        for (String path : examplePaths) {
+            final PathAndQuery pathAndQuery = PathAndQuery.parse(path);
+            checkArgument(pathAndQuery != null, "examplePaths contains an invalid path: {}", path);
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/docs/MethodInfo.java]: `{}` -> `%s`

File: core/src/main/java/com/linecorp/armeria/server/docs/MethodInfo.java [hunk 12]
```diff
@@ -117,16 +118,25 @@ public MethodInfo(String name,
...
-        this.exampleQueries.forEach(query -> {
+        requireNonNull(exampleQueries, "exampleQueries");
+        final ImmutableList.Builder<String> exampleQueriesBuilder =
+                ImmutableList.builderWithExpectedSize(Iterables.size(exampleQueries));
+        for (String query : exampleQueries) {
             final PathAndQuery pathAndQuery = PathAndQuery.parse(query.startsWith("?") ? query : '?' + query);
-            checkArgument(pathAndQuery != null,
-                          "exampleQueries contains an invalid query string: {}", query);
-        });
+            checkArgument(pathAndQuery != null, "exampleQueries contains an invalid query string: {}", query);
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/docs/MethodInfo.java]: `{}` -> `%s`

File: core/src/main/java/com/linecorp/armeria/server/docs/MethodInfo.java (hunk 11 above)
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/server/docs/MethodInfo.java]: Oops...

**Pull request merged**